src/utils/daos_serialize_hdlr.c

```diff
@@ -0,0 +1,281 @@
+/**
+ * (C) Copyright 2016-2021 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+
+/*
+ * daos_serialize_hdlr.c - handler function for daos serialization
+ * invoked by daos(8) utility
+ */
+
+#define D_LOGFAC	DD_FAC(client)
+#define LIBSERIALIZE	"libdaos_serialize.so"
+
+#include <stdio.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <dlfcn.h>
+#include <daos.h>
+#include <daos/common.h>
+#include <daos/debug.h>
+
+#include "daos_types.h"
+#include "daos_fs.h"
+#include "daos_uns.h"
+#include "daos_hdlr.h"
+#include "daos_datamover.h"
+
+/* calls daos_cont_serialize using dlsym inside of daos_serialize library */
+static int
+serialize_cont(struct cmd_args_s *ap, daos_prop_t *props, struct dm_stats *stats,
+	       struct dm_args *ca, char *filename)
+{
+	int		rc = 0;
+	void		*handle;
+	int		num_attrs = 0;
+	char		**names = NULL;
+	void		**buffers = NULL;
+	size_t		*sizes = NULL;
+	int (*daos_cont_serialize)(daos_prop_t *, int, char **, char **, size_t *, int *,
+				   int *, int *, uint64_t *, daos_handle_t, char *);
+
+	/* Get all user attributes if any exist */
+	rc = dm_cont_get_usr_attrs(ap, ca->src_coh, &num_attrs, &names, &buffers, &sizes);
+	if (rc != 0) {
+		DH_PERROR_DER(ap, rc, "Failed to get user attributes");
+		D_GOTO(out, rc);
+	}
+
+	handle = dlopen(LIBSERIALIZE, RTLD_NOW);
+	if (handle == NULL) {
+		rc = -DER_INVAL;
+		DH_PERROR_DER(ap, rc, "libdaos_serialize.so not found");
+		D_GOTO(out, rc);
+	}
+	daos_cont_serialize = dlsym(handle, "daos_cont_serialize");
+	if (daos_cont_serialize == NULL)  {
+		rc = -DER_INVAL;
+		DH_PERROR_DER(ap, rc, "Failed to lookup daos_cont_serialize");
+		D_GOTO(out, rc);
+	}
+	(*daos_cont_serialize)(props, num_attrs, names, (char **)buffers, sizes, &stats->total_oids,
+			       &stats->total_dkeys, &stats->total_akeys, &stats->bytes_read,
+			       ca->src_coh, filename);
+out:
+	if (num_attrs > 0)
+		dm_cont_free_usr_attrs(num_attrs, &names, &buffers, &sizes);
+	return rc;
+}
+
+int
+cont_serialize_hdlr(struct cmd_args_s *ap)
+{
+	int			rc = 0;
+	int			rc2 = 0;
+	char			*src_str = NULL;
+	size_t			src_str_len = 0;
+	daos_cont_info_t	src_cont_info = {0};
+	struct dm_args		ca = {0};
+	struct dm_stats		stats = {0};
+	size_t			size = 0;
+	char			*filename = NULL;
+	daos_prop_t		*props = NULL;
+
+	/* if output path is not passed in use current working dir */
+	if (ap->output_path == NULL) {
+		ap->output_path = getcwd(ap->output_path, size);
+		if (ap->output_path == NULL)
+			D_GOTO(out, rc = -DER_NOMEM);
+	} else {
+		rc = mkdir(ap->output_path, S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);
+		if (rc != 0 && errno != EEXIST) {
+			DH_PERROR_DER(ap, rc, "Failed to create output directory");
+			D_GOTO(out, rc);
```

Comment:
@daltonbohning I disagree because cont deserialize should absolutely only take one file (unlike the parallel version which creates multiple). Serialize takes an optional directory because it creates the file. I thought about this some but if that directory happens to have more than one file in it then it is unclear/messy which file to use.

Reply:
@daltonbohning like if an output dir wasn't used on serialize and it is written to the current working directory, then someone passed in the current working directory to deserialize... you could maybe guess it was the file with .h5? But I think that feels error prone. instead of just specifying the exact file to deserialize. The only reason I went with doing it that way in the parallel version was because there could be so many files and a real pain to specify them all on the command line, but this is just one file.

Reply:
@daltonbohning maybe if the names of the files were serialized into the hdf5 file it could just look for those on deserialization. This would require a non-trivial update to all tools though.

Reply:
@dsikich I just find this flow strange, because serialize and deserialize have disjoint interfaces. I.e. the user can't specify the filename in serialize, but has to supply it in deserialize.
```
# Serialize takes a DIR
daos cont serialize -o some_dir
# some_dir/file.h5 created automatically

# Deserialize takes a FILE
daos cont deserialize some_dir/file.h5
```
With pool and container creation, this problem is solved by having a --label option so the user knows the reference. For these tools, maybe the simple approach is for the user to just supply the FILEname instead of directory? Maybe someone else has another opinion

Reply:
@daltonbohning the user doesn't supply a name in serialize because it is created by the tool, so the same file created by tool is passed to deserialize. I don't know maybe it only makes sense to me, but having the user specify the filename in serialize would probably be fine if assumed they pass that to deserialize.